src/query_builder.rs

```diff
@@ -0,0 +1,737 @@
+        self.get_value_for_var(var).map_or(None, |v| v.as_kw())
+    }
+
+    fn var_as_boolean(&self, var: &str) -> Option<bool> {
+        self.get_value_for_var(var).map_or(None, |v| v.as_boolean())
+    }
+
+    fn var_as_long(&self, var: &str) -> Option<i64> {
+        self.get_value_for_var(var).map_or(None, |v| v.as_long())
+    }
+
+    fn var_as_double(&self, var: &str) -> Option<f64> {
+        self.get_value_for_var(var).map_or(None, |v| v.as_double())
+    }
+
+    fn var_as_instant(&self, var: &str) -> Option<DateTime<Utc>> {
+        self.get_value_for_var(var).map_or(None, |v| v.as_instant())
+    }
+
+    fn var_as_timestamp(&self, var: &str) -> Option<i64> {
+        self.get_value_for_var(var).map_or(None, |v| v.as_timestamp())
+    }
+
+    fn var_as_string(&self, var: &str) -> Option<Rc<String>> {
+        self.get_value_for_var(var).map_or(None, |v| v.as_string())
+    }
+
+    fn var_as_uuid(&self, var: &str) -> Option<Uuid> {
+        self.get_value_for_var(var).map_or(None, |v| v.as_uuid())
+    }
+
+    fn var_as_uuid_str(&self, var: &str) -> Option<String> {
+        self.get_value_for_var(var).map_or(None, |v| v.as_uuid_string())
+    }
+}
+
+impl ColResultsRow for TupleResult {
+    fn col_as_entid(&self, col: usize) -> Option<Entid> {
+        self.get_value_for_column(col).map_or(None, |v| v.as_entid())
+    }
+
+    fn col_as_kw(&self, col: usize) -> Option<Rc<NamespacedKeyword>> {
+        self.get_value_for_column(col).map_or(None, |v| v.as_kw())
+    }
+
+    fn col_as_boolean(&self, col: usize) -> Option<bool> {
+        self.get_value_for_column(col).map_or(None, |v| v.as_boolean())
+    }
+
+    fn col_as_long(&self, col: usize) -> Option<i64> {
+        self.get_value_for_column(col).map_or(None, |v| v.as_long())
+    }
+
+    fn col_as_double(&self, col: usize) -> Option<f64> {
+        self.get_value_for_column(col).map_or(None, |v| v.as_double())
+    }
+
+    fn col_as_instant(&self, col: usize) -> Option<DateTime<Utc>> {
+        self.get_value_for_column(col).map_or(None, |v| v.as_instant())
+    }
+
+    fn col_as_timestamp(&self, col: usize) -> Option<i64> {
+        self.get_value_for_column(col).map_or(None, |v| v.as_timestamp())
+    }
+
+    fn col_as_string(&self, col: usize) -> Option<Rc<String>> {
+        self.get_value_for_column(col).map_or(None, |v| v.as_string())
+    }
+
+    fn col_as_uuid(&self, col: usize) -> Option<Uuid> {
+        self.get_value_for_column(col).map_or(None, |v| v.as_uuid())
+    }
+
+    fn col_as_uuid_string(&self, col: usize) -> Option<String> {
+        self.get_value_for_column(col).map_or(None, |v| v.as_uuid_string())
+    }
+}
+
+impl IntoIterator for TupleResult {
+    type Item = (String, TypedValue);
+    type IntoIter = indexmap::map::IntoIter<String, TypedValue>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.map.into_iter()
+    }
+}
+
+pub struct RelResult {
+    rows: Vec<TupleResult>,
```

Comment:
This is expensive. You're heap-allocating a new map for each row, and doing so eagerly on creation, even if the caller only wants to get the first three rows. (It's bad enough that our rel results are `Vec<Vec<TypedValue>>`; they should really be `Vec<TypedValue>` with a width.)

Does `RelResult` buy us anything over using the raw result?